Add ConfigStringMap

kr1viah opened the pull request:
Related issue: https://github.com/sakura-ryoko/malilib/issues/117

kr1viah committed:
ConfigStringMap

kr1viah committed:
Add missing constructors

kr1viah committed:
Fix things not updating if only one (value or key) has been updated

kr1viah commented:
Actually, don't merge yet; I have some ideas to improve this. being labels, and double/float/integer maps (like string to float)

kr1viah commented:
I think I'll make a builder instead of the usual 5-6 constructors considering the amount of arguments (Note: only name and types are required)
```java
public ConfigTable(String name, String comment, String prettyName, String translatedName,
                   @Nullable String displayString, List<List<Object>> defaultValue,
                   List<String> labels, boolean showEntryNumbers, boolean allowAddNewEntry,
                   Class<?>... types)
{```

sakura-ryoko commented:
Just a suggestion instead of using the `Class<?>` argument; try using the `Codec<T>` interface and have it be class type bound with `T` instead of `Object`; and if you need more than 1 type, use two Generics; etc.

kr1viah committed:
GuiTableEdit

kr1viah commented:
All done. I'm not sure if the CODEC is correct (I haven't tested), because I don't know how to test it.

kr1viah commented:
> Just a suggestion instead of using the `Class<?>` argument; try using the `Codec<T>` interface and have it be class type bound with `T` instead of `Object`; and if you need more than 1 type, use two Generics; etc.

I'll see if I can.

Edit: I can't. My fault for being terrible at naming I guess. `types` in the constructor there is used for defining which type each horizontal entry in the table should be. For example, if you set types to `Integer.class, Integer.class` you'll get a table with horizontally 2 integer text widgets (You know, a GuiTextFieldInteger), and if you do `String.class, Double.class, Integer.class` you'll get a table with horizontally a normal text widget, a double text widget and an integer text widget.

kr1viah committed:
Change constructor to private, setEntryCount()
setEntryCount() prefills defaultValue with either the first entry of defaultValue (if there is only one entry) or with dummy's

kr1viah committed:
Better usage experience

kr1viah committed:
Better usage experience

kr1viah committed:
Fix various crashes, test configs, Entry.equals()

sakura-ryoko commented:
I see you found my Test Configs.  Please don't forget to add the i18n (en_us only) for them.  If you haven't figured it out yet it needs `MaLiLibReference.LOCAL_DEBUG` set to true to enable.

It's exactly for this sort of thing; testing how it looks and functions in game.

kr1viah committed:
Make methods that should be public public, improved test configs, fix ConfigTable#build()

kr1viah commented:
> it needs MaLiLibReference.LOCAL_DEBUG set to true to enable.

I just forced MaLiLibConfigGui to load them, and I'm just not committing that. Perhaps make a program argument for that? 

> Please don't forget to add the i18n (en_us only) for them

Will do now.

sakura-ryoko commented:
> > it needs MaLiLibReference.LOCAL_DEBUG set to true to enable.
> 
> I just forced MaLiLibConfigGui to load them, and I'm just not committing that. Perhaps make a program argument for that?
> 

LOCAL_DEBUG is the argument.  In order for all of the Debug functions to be activated it needs to be set to TRUE at Java init.

Just don't leave it turned ON when submitting the code.  I am not fond of appending Java Program Arguments to Minecraft unless I have to.

kr1viah committed:
Language, todo

kr1viah committed:
Fix slight overlap in GuiTableEdit

kr1viah commented:
Okay. *Now* it is done I think.

Coincidentally this also removes the need for a 2d/3d position config, because with this you can make a table of height one, with 2/3 integer/double entries, and set showEntryNumbers and allowAddNewEntry both to false.

Minor nitpick: with GuiTextFieldInteger and GuiTextFieldDouble you can't ctrl + backspace, because an empty input is considered invalid. I think an empty input should be turned into 0, like how 024 gets turned into 24. Should I fix/do/change this?

kr1viah committed:
Remove duplicate code

sakura-ryoko commented:
There is a value validator for those text boxes.  Here -->
https://github.com/sakura-ryoko/malilib/blob/eaa0fa425558b0218a1f37bb2ec01f44658e6fa5/src/main/java/fi/dy/masa/malilib/gui/GuiTextFieldDouble.java#L21

Same with the Integer box.

sakura-ryoko commented:
You can say it was a TO DO thing; it is mostly used in MiniHUD's Shape Editor and LM; so not many places to test it within MaLiLib.

kr1viah commented:
Something in me feels like also adding support for keybinds and booleans and dropdowns and whatnot, but the rest of me tells me that that is way overkill haha.

kr1viah commented:
It's too tempting. I'm gonna do it.

sakura-ryoko commented:
> It's too tempting. I'm gonna do it.

You are just adding a few additional config types; not trying to re-invent the config system.

Remember that.

kr1viah commented:
> You are just adding a few additional config types

Errmm... It's actually one config type. It just happens to be very versatile lol.

kr1viah committed:
Rework using separate classes
I could spend time naming this commit better but eh

kr1viah committed:
Update json en/decoding

kr1viah committed:
Just gui left

kr1viah committed:
I've run into the same bug 3 times in a row and still don't know how I fixed it last time

kr1viah committed:
I fixed that bug

kr1viah committed:
Remove useless anonymous object

kr1viah committed:
Keybinds now actually save and update

kr1viah commented:
I think keybinds in tables now work, though I'm not sure how to make the keybinds actually do something.

I'll add booleans, which should be pretty simple, and after that I don't think there is another type that could be added.

kr1viah committed:
Missing entry

kr1viah committed:
BooleanEntry, fix reset button being enabled when it shouldn't be.

kr1viah commented:
> I'm not sure how to make the keybinds actually do something.

Still pondering about this.

kr1viah committed:
Wider dialog

kr1viah commented:
I am still not sure how to make the keybinds do something.

One thing that makes this a bit difficult, is from my testing if you register a keybind while in game it won't do anything until you rejoin.

sakura-ryoko commented:
FYI, you'll need to update your branch.  MaLiLib was converted for 1.21.11+

kr1viah commented:
Alright I've decided. I won't remove keybind/hotkey (I still don't know what to call them) entries, considering they do work as intended, but I won't make them do anything either.

How should I go about updating my branch? My code barely uses any Minecraft related classes (just `Codecs`, `Click`, `DrawContext`, `Screen`, `KeyInput`, `CharInput` and `Pair`), so the updating itself should be easy.

sakura-ryoko commented:
Yeah I'd like to get this wrapped up soon because I will need to convert the entire mods to Mojmap very soon.

sakura-ryoko commented:
I don't think it having Keybinds in some config-map makes any sense; personally.

kr1viah commented:
Alright, I'll remove the keybinds and after that it's done.

kr1viah commented:
Should I bother changing the formatting in my code to match the rest?

sakura-ryoko commented:
> Should I bother changing the formatting in my code to match the rest?

Yes please keep it all consistent.

kr1viah committed:
Remove KeybindEntry

kr1viah committed:
Missing type

kr1viah committed:
Formatting

kr1viah commented:
Done now. I probably missed something with the formatting, though I've checked a couple of times and found nothing.

sakura-ryoko commented:
I'll merge this and experiment in Snapshot; not for 1.21.10.  With 1.21.10; that ship has already sailed for such a change.

sakura-ryoko commented:
I'll clean up more of it in DEV mode.
